In [1]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

Print function ready to serve.


In [3]:
import math

# total Volume of water you can carry
total_volume = math.pi*math.pow(5,2) * 17
total_volume

1335.1768777756622

In [4]:
# do you have space for 2000cm^3
if total_volume >= 2000:
    print(True)
else:
    print(False)    

False


In [5]:
bottles = total_volume/5000
bottles

0.26703537555513246

In [15]:
#print message to Q
"Yes, i have enough space. I can carry a total-volume cm^3 of water"
print (("No, I don't have enough space. I can carry"), str(total_volume),("cm^3 of water"))

No, I don't have enough space. I can carry 1335.1768777756622 cm^3 of water


In [17]:
bottles =3
bottle_height =16
bottle_radius =4

In [19]:
#Volume of bottle
bottle_volume = math.pi*bottle_radius**2*bottle_height
#print the volume of a single bottle
print (bottle_volume)

804.247719318987


In [21]:
#Total Volume
total_volume = bottles* bottle_volume
#print the Total Volume
print(total_volume)

2412.743157956961


In [28]:
from __future__ import print_function
print("Print function ready to serve")

Print function ready to serve


i love this exercise and i am really enjoying so much

In [2]:
# Read lists of locations (simply run this code block)
with open('project_1files/brent.txt', 'r') as f:
    brent = f.read().splitlines()
    
with open('project_1files/camden.txt', 'r') as f:
    camden = f.read().splitlines()

with open('project_1files/redbridge.txt', 'r') as f:
    redbridge = f.read().splitlines()

with open('project_1files/southwark.txt', 'r') as f:
    southwark = f.read().splitlines()

In [4]:
print(type(brent) )

<class 'list'>


In [4]:
#print the first five locations in Camden
camden[0:5] 

['Chalk Farm', 'Dartmouth Park', 'Fitzrovia', 'Frognal', 'Bloomsbury']

In [3]:
# print the length of each list
c_length =len(camden)
b_length =len(brent)
r_length =len(redbridge)
s_length =len(southwark)

c_length == max(c_length, b_length, r_length, s_length)
print("Camden has the most locations.")


Camden has the most locations.


In [4]:
# is 'Newbury Park' in Redbridge?
if "Newbury Park" in redbridge:
    print (True)
else:
    print(False)
    
# is 'Peckam' in Brent?
if "Peckham" in brent:
    print(True)
else:
    print (False)
    


True
False


In [5]:
# Print minimum value in southwark
sorted(southwark)[0]
# Print maximum value in southwark
sorted(southwark)[-1]

#The list are ordered alphabetically


'Walworth'

In [6]:
# Brent has duplicates?
b_len =len(brent)
print(b_len>len(set(brent)))

# Camden has duplicates?
c_len =len(camden)
print(c_len>len(set(camden)))
# Redbridge has duplicates?
r_len =len(redbridge)
print(r_len>len(set(redbridge)))
# Southwark has duplicates?
s_len =len(southwark)
print(s_len>len(set(southwark)))

False
True
True
True


In [8]:
camden = list(set(camden))
redbridge =list(set(redbridge))
southwark = list(set(southwark))

In [9]:
# Create location_dict
location_dict ={"Brent":brent,"Camden": camden,"Redbridge":redbridge,"Southwark":southwark}

In [10]:
# Run this cell
for borough in ['Brent','Camden','Redbridge','Southwark']:
    print(borough in location_dict)
    

True
True
True
True


In [23]:
#Run this cell
brent, camden, redbridge, southwark =None, None, None, None


In [24]:
print(type(None))

<class 'NoneType'>


In [11]:
#Create a new key-value pair for 'Inner London'
location_dict["Inner London"]= location_dict["Camden"] + location_dict["Southwark"]

#Create a new key-value pair for 'Outer london'
location_dict["Outer London"]= location_dict["Brent"]+location_dict["Redbridge"]

In [12]:
print(len(location_dict['Inner London']))
print(len(location_dict['Outer London']))


32
35


In [13]:
import pickle

#Save object to disk
with open('location_dict.pkl','wb') as f:
    pickle.dump(location_dict, f)

In [14]:
from __future__ import print_function  #Compatibility with Python 3
print('Print function ready to serve')

Print function ready to serve


In [32]:
import pickle

#Read object from disk
with open('location_dict.pkl','rb')as f:
    location_dict =pickle.load(f)

In [15]:
#Print the keys in the locatio_dict
for keys in location_dict.keys():
    print(keys)
    

Brent
Camden
Redbridge
Southwark
Inner London
Outer London


In [37]:
#Write code here
if len(location_dict["Inner London"]) > len(location_dict["Outer London"]):
    print("I want to start in Inner London")
elif len(location_dict["Outer London"]) > len(location_dict["Inner London"]):
    print ("I want to start in Outer London")
else:
    print (" Either is fine. flip a coin, Q.")

I want to start in Outer London


In [16]:
# For each key in location_dict, print the number of locations in its list
for key, value in location_dict.items(): 
    print ("{} has {} locations".format(key, len(value)))

Brent has 19 locations
Camden has 19 locations
Redbridge has 16 locations
Southwark has 13 locations
Inner London has 32 locations
Outer London has 35 locations


Now, let's give each location in Outer London a first impression based on its name. 

<br>
**Combine <code style="color:steelblue">if</code> and <code style="color:steelblue">for</code> statements. For each location in Outer London...**
* **If** its name has <code style="color:steelblue">'Farm'</code>, <code style="color:steelblue">'Park'</code>, <code style="color:steelblue">'Hill'</code>, or <code style="color:steelblue">'Green'</code> in it, print:

<pre style="color:steelblue">{<strong>name</strong>} sounds pleasant.</pre>

* **Else If** its name has <code style="color:steelblue">'Royal'</code>, <code style="color:steelblue">'Queen'</code>, or <code style="color:steelblue">'King'</code> in it, print:

<pre style="color:steelblue">{<strong>name</strong>} sounds grand.</pre>
    
* If its name doesn't sound pleasant or grand, just ignore it.
* **Tip:** If you want to check if any word from a list is found in a string, you can use <code style="color:steelblue">any()</code>, like so:

<pre style="color:steelblue">
any( word in name for word in <strong>list_of_words</strong> )
</pre>


In [17]:
pleasant_sounding = ['Farm', 'Park', 'Hill', 'Green']
royal_sounding = ['Royal', 'Queen', 'King']

# Print first impression of each location in Outer London based on its names
for locations in location_dict["Outer London"]:
    if any(word in locations for word in pleasant_sounding):
        print(locations +  " sounds pleasant.")
    elif any(word in locations for word in royal_sounding):
        print (locations + " sounds grand.")

Brent Park sounds pleasant.
Dollis Hill sounds pleasant.
Kensal Green sounds pleasant.
Kingsbury sounds grand.
Queen's Park sounds pleasant.
Wembley Park sounds pleasant.
Park Royal sounds pleasant.
Woodford Green sounds pleasant.
Newbury Park sounds pleasant.
Gants Hill sounds pleasant.
Seven Kings sounds grand.


In [18]:
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 3.3</span>

**Using a list comprehension, create a new list called <code style="color:steelblue">pleasant_locations</code>.**
* It should contain locations in Outer London that are <code style="color:steelblue">pleasant_sounding</code>.
* Then print the list.
* **Tip:** To check if any word from a list is found in a string, you can use <code style="color:steelblue">any()</code>. See <span style="color:RoyalBlue">Exercise 4.2</span> if you need a reminder.

SyntaxError: invalid syntax (<ipython-input-18-4b01e15bcbeb>, line 1)

In [19]:
## Create pleasant_locations list using a list comprehension
pleasant_locations = [locations for locations in location_dict["Outer London"] if any(word in locations for word in pleasant_sounding)]
# Print the pleasant-sounding locations
print (pleasant_locations) 

['Brent Park', 'Dollis Hill', 'Kensal Green', "Queen's Park", 'Wembley Park', 'Park Royal', 'Woodford Green', 'Newbury Park', 'Gants Hill']


In [21]:
##Print the number pleasant-sounding locations we have.**

In [22]:
# Print number of pleasant-sounding locations
print (str(len(pleasant_locations)))

9


In [23]:
# Code here
def filter_locations(location_list, word_list):
    names = [locations for locations in location_list if any(word in locations for word in word_list)]
    return names

In [24]:
# Create pleasant_locations using filter_locations()
pleasant_locations = filter_locations(location_dict["Outer London"], pleasant_sounding)

# Print list of pleasant-sounding locations
print (pleasant_locations)


['Brent Park', 'Dollis Hill', 'Kensal Green', "Queen's Park", 'Wembley Park', 'Park Royal', 'Woodford Green', 'Newbury Park', 'Gants Hill']


In [25]:
**Next, let's use this handy function to create a <code style="color:steelblue">grand_locations</code> list for locations that sound grand.**
* Pass in the list of Outer London locations and the list of grand-sounding words.
* Print the new list and confirm the expected output

SyntaxError: invalid syntax (<ipython-input-25-67b486e61a56>, line 1)

In [26]:
# Create grand_locations using filter_locations()
grand_locations = filter_locations(location_dict["Outer London"], royal_sounding)

# Print list of grand-sounding locations
print (grand_locations)


['Kingsbury', "Queen's Park", 'Park Royal', 'Seven Kings']


In [27]:
# Display locations that sound pleasant and grand
print (set(pleasant_locations).intersection(set(grand_locations)))

{'Park Royal', "Queen's Park"}


In [28]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

Print function ready to serve.


In [42]:
import numpy as np

In [43]:
#Create array of business_ids with values ranging from 1 to 1701
business_ids =np.arange(1, 1701)

In [44]:
# Print shape of business_ids 
print(business_ids.shape)

(1700,)


In [45]:
# Print last 10 business ID's
print(business_ids[-11:])

[1690 1691 1692 1693 1694 1695 1696 1697 1698 1699 1700]


In [46]:
# Create id_matrix by reshaping business_ids to have 10 columns
id_matrix = np.reshape(business_ids, (170, 10))

# Print shape
print (id_matrix)

[[   1    2    3 ...    8    9   10]
 [  11   12   13 ...   18   19   20]
 [  21   22   23 ...   28   29   30]
 ...
 [1671 1672 1673 ... 1678 1679 1680]
 [1681 1682 1683 ... 1688 1689 1690]
 [1691 1692 1693 ... 1698 1699 1700]]


In [47]:
# Reshape business_ids to have 10 rows, with 170 businesses each

id_matrix = id_matrix.reshape(10, 170)

In [48]:
# Print the first row of id_matrix
print(id_matrix[0,:])


[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170]


In [49]:
# Overwrite id_matrix with flipped version
id_matrix = np.transpose(id_matrix)

# Print first column
print(id_matrix[:,0])


# Print shape of new id_matrix
print (id_matrix.shape)


[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170]
(170, 10)


In [52]:
# Overwrite id_matrix with flipped version
id_matrix = np.transpose(id_matrix)

# Print first column
print(id_matrix[:,0])


# Print shape of new id_matrix
print (id_matrix.shape)


[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170]
(170, 10)


In [53]:
# Create group_min
group_min = np.min(id_matrix, axis=0)
# Print group_min
print (group_min)


[   1  171  341  511  681  851 1021 1191 1361 1531]


In [54]:
# Create group_max
group_max = np.max(id_matrix, axis=0)
# Print group_max
print (group_max)

[ 170  340  510  680  850 1020 1190 1360 1530 1700]


In [55]:
# Print range of each group
print (group_max - group_min + 1)

[170 170 170 170 170 170 170 170 170 170]


In [58]:
# Set random seed
np.random.seed(9001)
# Print selected businesses from each group
for group in range(id_matrix.shape[1]): 
    print("Group {}: {}".format(group+1, np.random.choice(id_matrix[:,group], 5, replace=False)))
    

Group 1: [  7  37 124  41  17]
Group 2: [302 261 257 323 234]
Group 3: [464 486 463 440 474]
Group 4: [645 582 666 631 553]
Group 5: [699 738 705 760 792]
Group 6: [ 995 1000  909  869  899]
Group 7: [1042 1058 1172 1122 1153]
Group 8: [1304 1275 1343 1344 1236]
Group 9: [1401 1417 1383 1387 1474]
Group 10: [1642 1687 1545 1644 1549]


In [59]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

Print function ready to serve.


In [1]:
import pandas as pd


Read Iris dataset:

In [3]:
# Read the iris dataset from a CSV file
df = pd.read_csv('project_1files/iris.csv')

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.1</span>

***In the previous lesson...***

> *In the previous lesson, we discovered that **Casey's Flower Shop** is the first local business we need to visit in Park Royal.*


When arrive, we find out that the owner, Casey, urgently needs our help!
> * They just received a new shipment of **Iris flowers**, but they've never stocked these flowers before.
* Casey asks us to share what we know about these new flowers.
* Luckily, we have the Iris dataset to learn more about them ourselves!

In Lesson 4: NumPy, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts.

While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

<br>
**First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.**
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [4]:
# Create toy_df
df_h=df.head()
df_t=df.tail()
toy_df=pd.concat([df_h, df_t])


**Next, display <code style="color:steelblue">toy_df</code>.** 
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [5]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

You should have 'setosa' and 'virginica' flowers.

**Next, display a summary table for <code style="color:steelblue">toy_df</code>.**
* It should have the mean, standard deviation, and quartiles for each of the columns

In [7]:
#Describe toy_df
toy_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count     10.000000    10.000000     10.000000    10.000000
mean       5.590000     3.130000      3.290000     1.130000
std        0.807534     0.316403      1.995244     0.992248
min        4.600000     2.500000      1.300000     0.200000
25%        4.925000     3.000000      1.400000     0.200000
50%        5.500000     3.050000      3.250000     1.000000
75%        6.275000     3.350000      5.175000     1.975000
max        6.700000     3.600000      5.400000     2.300000

Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.2</span>

Elementwise operations are very useful in machine learning, especially for feature engineering.

<br>

> *<span style="color:tomato; font-weight:bold">Feature engineering</span> is the process of creating new features (model input variables) from existing ones.*

<br>

We'll cover this topic in much more detail in Projects 2, 3, and 4, but let's first use our <code style="color:steelblue">toy_df</code> to illustrate the concept.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

<br>
**First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.**
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [10]:
#Display petal_width and petal_length
petal_width =toy_df["petal_width"]
petal_length =toy_df["petal_length"]
toy_df[["petal_width","petal_length"]]

petal_width  petal_length
0            0.2           1.4
1            0.2           1.4
2            0.2           1.3
3            0.2           1.5
4            0.2           1.4
145          2.3           5.2
146          1.9           5.0
147          2.0           5.2
148          2.3           5.4
149          1.8           5.1

**Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.**
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [11]:
# Create a new petal_area column
toy_df["petal_area"] = toy_df["petal_width"] * toy_df["petal_length"]

# Display toy_df
toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

**Finally, what do we now know about Iris flowers?**

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do!

Often, by creating new features, you can learn more about the data (and improve your machine learning models).

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.3</span>

Again, we'll use the <code style="color:steelblue">toy_df</code> to really drive home these concepts.

Let's say we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

<br>
**First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.**

In [12]:
#Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

**Take a look at the DataFrame and manually count the number that satisfy our conditions.**
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

Great. Now we'll see what's going on under the hood when we use our boolean masks.

<br>
**Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.**
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [15]:
#Mask for petal_area >10

petal_area_mask=toy_df["petal_area"]>10
#Display petal_area_mask
petal_area_mask

0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool

**Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.**
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [16]:
# Mask for sepal_width > 3
sepal_width_mask = toy_df["sepal_width"] > 3

# Display sepal_width_mask
sepal_width_mask



0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool

**Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.**
* Note how their combination results in another boolean mask!

In [17]:
# Display both masks, combined
sepal_width_mask & petal_area_mask


0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool

**Finally, select the observations from <code style="color:steelblue">toy_df</code> where both conditions are met.**

In [18]:
# Index with both masks

toy_df[sepal_width_mask & petal_area_mask]

sepal_length  sepal_width  petal_length  petal_width    species  \
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
148       12.42

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.4</span>

Now, armed with the power of **groupby**, we're almost ready to return to Casey and share what we've learned about Iris flowers! But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

<br>
Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. 
* Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.

<br>
**First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.**
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [19]:
# Display all 'virginica' species, sorted by petal_area
toy_df[toy_df["species"] == "virginica"].sort_values("petal_area", axis=0, ascending="True")


sepal_length  sepal_width  petal_length  petal_width    species  \
149           5.9          3.0           5.1          1.8  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
145           6.7          3.0           5.2          2.3  virginica   
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
149        9.18  
146        9.50  
147       10.40  
145       11.96  
148       12.42

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

<br>
**Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.**
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.

In [20]:
# Display all 'setosa' species
toy_df[toy_df["species"] == "setosa"].sort_values("petal_area", axis=0, ascending="True")


sepal_length  sepal_width  petal_length  petal_width species  petal_area
2           4.7          3.2           1.3          0.2  setosa        0.26
0           5.1          3.5           1.4          0.2  setosa        0.28
1           4.9          3.0           1.4          0.2  setosa        0.28
4           5.0          3.6           1.4          0.2  setosa        0.28
3           4.6          3.1           1.5          0.2  setosa        0.30

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

<br>
**Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.**
* Do you get the same result?

In [21]:
# Median petal_area in toy_df
toy_df.groupby("species")["petal_area"].median()

species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64

** *Congratulations... You've completed the training mission!* **

> *In this lesson, you explored the Iris dataset using Pandas.*

> *After arming yourself with Pandas, you've successfully completed the rest of the training mission!*

<br>
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>